[KIC 2.0] Add knative Ingress support

ccfishk opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md and ensure your changes are being reflected in CHANGELOG.md for the next upcoming release
-->

**What this PR does / why we need it**:
what:
  Using KIC 2.0 with Knative 

Why:
  Knative is supported in KIC 1.3.0. Need migrate it to KIC 2.0

**Checklist**
Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR
- [ ]  Continuation Test including dbless, postgress and legacy all pass.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/kubernetes-ingress-controller?pullRequest=1396) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/kubernetes-ingress-controller?pullRequest=1396) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/kubernetes-ingress-controller?pullRequest=1396) it.</sub>

shaneutt commented on README.md:
```diff
@@ -137,3 +137,5 @@ meeting invite and Zoom links to join the meeting.
 [kong-logo]: https://konghq.com/wp-content/uploads/2018/05/kong-logo-github-readme.png
 [k4k8s-enterprise-setup]: https://docs.konghq.com/kubernetes-ingress-controller/latest/deployment/k4k8s-enterprise/
 [guides]: https://docs.konghq.com/kubernetes-ingress-controller/latest/guides/overview/
+
+### migrate knative ingress supoort from v1.3.x to v2.0. 1
```
It would appear we intended to add some documentation here which covers the effects on knative.Ingress when upgrading to v2.0, we need to finish this documentation (or remove it if it's no longer relevant).

ccfishk replied:
I have been understanding this concise documentation should be moved to Changelog.md ?

shaneutt replied:
Sure, but we also seem to need to expand upon it because what I'm seeing currently just looks like a header, no?

ccfishk replied:
A clear documentation (what is added) should be able to satisfy the need.

shaneutt commented on internal/ingress/status/status.go:
```diff
@@ -559,6 +559,7 @@ func (s *statusSync) runUpdateTCPIngress(ctx context.Context,
 			return nil, nil
 		}
 
+		fmt.Println("runUpdateTCPIngress heheheh")
```
This looks like a debugging line we need to remove:

```suggestion
```

ccfishk replied:
Good cache.

shaneutt commented on pkg/parser/translate.go:
```diff
@@ -441,15 +441,13 @@ func fromUDPIngressV1Alpha1(log logrus.FieldLogger, ingressList []*v1alpha1.UDPI
 }
 
 func fromKnativeIngress(log logrus.FieldLogger, ingressList []*knative.Ingress) ingressRules {
-
 	sort.SliceStable(ingressList, func(i, j int) bool {
 		return ingressList[i].CreationTimestamp.Before(
 			&ingressList[j].CreationTimestamp)
 	})
 
 	services := map[string]kongstate.Service{}
 	secretToSNIs := newSecretNameToSNIs()
-
```
Generally speaking I would avoid making whitespace changes in files that are not otherwise going to get any updates as a part of this PR, I would suggest we leave this file alone unless we're changing something substantive.

ccfishk replied:
Agree.

shaneutt commented on pkg/store/store.go:
```diff
@@ -424,16 +424,21 @@ func (s Store) validKnativeIngressClass(objectMeta *metav1.ObjectMeta) bool {
+		s.stores.KnativeIngress,
+		labels.NewSelector(),
 		func(ob interface{}) {
+			fmt.Printf("\nwithin ListKnativeIngresses cache.ListAll")
 			ing, ok := ob.(*knative.Ingress)
+			fmt.Printf("\nwithin ListKnativeIngresses checking ing %v", ing)
```
These look like debug prints that we need to make sure we clean up.

ccfishk replied:
my bad. thanks.

shaneutt commented on railgun/apis/addtoscheme_knative_v1alpha1.go:
```diff
@@ -0,0 +1,25 @@
+
+func init() {
+	// Register the types with the Scheme so the components can map objects to GroupVersionKinds and back
+	AddToSchemes = append(AddToSchemes, v1alpha1.SchemeBuilder.AddToScheme)
```
It doesn't seem like we would need to add add knative.Ingress to the runtime.Scheme on package load, but instead the controller manager would need to add it prior to starting the controller if it's enabled. I would suggest we at least look into an alternative approach here where we only add knative to the scheme opportunistically if its enabled.

ccfishk replied:
Would like to follow good strategy - apply runtime scheme opportunistically.

shaneutt replied:
In that case are you on board with dropping these init() scheme loaders in favor of loading only when enabled in the manager?

ccfishk replied:
Answer your question in another similar thread.

shaneutt commented on railgun/apis/apis.go:
```diff
@@ -0,0 +1,27 @@
+/*
+Copyright 2021 Kong, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apis
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var AddToSchemes runtime.SchemeBuilder
+
+func AddToScheme(s *runtime.Scheme) error {
+	return AddToSchemes.AddToScheme(s)
+}
```
Similar to my above comments: it would seem preferable to manage this in the controller manager.

ccfishk replied:
Similar answer as another comment.

shaneutt commented on railgun/hack/generators/controllers/networking/main.go:
```diff
@@ -20,9 +20,10 @@ const (
+	kongv1          = "github.com/kong/kubernetes-ingress-controller/railgun/apis/configuration/v1"
+	kongv1alpha1    = "github.com/kong/kubernetes-ingress-controller/railgun/apis/configuration/v1alpha1"
+	kongv1beta1     = "github.com/kong/kubernetes-ingress-controller/railgun/api/configuration/v1beta1"
+	knativev1alpha1 = "knative.dev/networking/pkg/apis/networking/v1alpha1"
```
We need to double check that `gofmt` actually would leave this import in this order, I think it will organize it _before_ the kong entries and we should reflect that in this template so that the generator is not producing files that `gofmt` and `goimports` will want to overwrite constantly.

ccfishk replied:
I believe it is pretty clear that a new refer is added.

shaneutt commented on railgun/internal/ctrlutils/vars.go:
```diff
@@ -6,3 +6,4 @@ package ctrlutils
 
 // KongIngressFinalizer is the finalizer used to ensure Kong configuration cleanup for deleted resources.
 const KongIngressFinalizer = "configuration.konghq.com/ingress"
+const KnativeIngressFinalizer = "networking.internal.knative.dev/ingress"
```
Can you help me to understand the impetus to add this finalizer? Generally speaking when a controller manages objects that belong to another API group (which themselves have their own controllers with finalizers) you would simply use a finalizer unique to the controller at hand such as the existing one. What led to adding this new one?

ccfishk replied:
That was added to debugging message flow between two controller. No kong code use it. remove it now.

shaneutt commented on railgun/manager/run.go:
```diff
@@ -57,6 +59,7 @@ func Run(ctx context.Context, c *config.Config) error {
 	utilruntime.Must(konghqcomv1.AddToScheme(scheme))
 	utilruntime.Must(configurationv1alpha1.AddToScheme(scheme))
 	utilruntime.Must(configurationv1beta1.AddToScheme(scheme))
+	utilruntime.Must(knativev1alpha1.AddToScheme(scheme))
```
Above I was talking about adding knative to the scheme only in the runtime and not during package load, this line would seem to do just that so I expect the other files can be dropped.

ccfishk replied:
Agree. should be able to cleanup.

shaneutt commented on railgun/test/integration/helloworldgo.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: serving.knative.dev/v1
+kind: Service
+metadata:
+  name: helloworld-go
+  namespace: default
+spec:
+  template:
+    spec:
+      containers:
+        - image: gcr.io/knative-samples/helloworld-go
+          env:
+            - name: TARGET
+              value: Go Sample v1
```
We should deploy this service using the upstream corev1 api rather than raw yaml in the tests, e.g.:

```
svc := corev1.Service{...}
```

ccfishk replied:
I would say third party test sample varies per release, whats most convenient fits our needs. 

But if you really prefer clientgo style, we can do it as an improvement.

shaneutt replied:
Given that all the tests so far use "clientgo style", I would like to make this improvement so that we're doing these kinds of things one consistent way.

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,286 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	knativeCrds = "https://github.com/knative/serving/releases/download/v0.13.0/serving-crds.yaml"
+	knativeCore = "https://github.com/knative/serving/releases/download/v0.13.0/serving-core.yaml"
+)
```
These lines wont necessarily age all that well, but I understand it's functional :thinking: 

Perhaps we can add knative deployment to the KTF with variable versioning, and provide toplevel defaults?

rainest replied:
Yeah, agreed, installation of Knative itself should probably live in KTF since it's a separate non-KIC app that resides in the cluster. KIC tests would then handle only creating the Knative Service resources they need for the test, or we'd add a stock general-purpose Knative Service in KTF similar to the other stuff in https://github.com/Kong/kubernetes-testing-framework/tree/main/pkg/generators/k8s. The latter may be more appropriate since Knative features would be tied to the Knative version, and that'd be determined by KTF.

The CRD/core install used here is what we currently suggest [in our docs](https://docs.konghq.com/kubernetes-ingress-controller/1.3.x/guides/using-kong-with-knative/#install-knative), but it's a bit dated at this point--Knative is now a number of versions beyond 0.13, and they've since introduced [an Operator](https://knative.dev/docs/install/knative-with-operators/) that's probably now the recommended way of managing Serving and CRD installation.

That said, our goal here is more to get existing 1.x Knative functionality into 2.x, so it may be prudent to just stick with the flat 0.13 manifests for now. 0.14 was the first version the Operator was available for, and it does look like the [0.14 core](https://github.com/knative/serving/releases/tag/v0.14.0) has sort-of breaking changes--they have a migration Job for the new StorageVersion, but we probably want to upgrade our static CRs once rather than installing the older versions and migrating them each time. I'd also expect that the first release of the Operator is probably fairly buggy, so we'd be better-served by going all the way to 0.23 if we want to migrate onto it.

ccfishk replied:
I have been understanding KTF mainly apply to test, test environment configuration, perf test etc. Integration tests have many unstable dependencies.

ccfishk replied:
In the long time, I think it is a good idea to move integration stuff into KTF.

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,286 @@
+	knativeCore = "https://github.com/knative/serving/releases/download/v0.13.0/serving-core.yaml"
+)
+
+func isKnativeReady(ctx context.Context, cluster kind.Cluster) bool {
```
I personally would like to see private helper functions for the tests moved to the bottom of the test file so that the tests themselves are prominent, let me know if that sounds OK to you?

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,286 @@
+)
+
+func isKnativeReady(ctx context.Context, cluster kind.Cluster) bool {
+	timeout := 1
+	for timeout < 60 {
```
Using Golang's native `time` library you can create a timeout that doesn't require you to do do specific 1 second sleep intervals, e.g.:

```
timeout := time.Now().Add(time.Minute * 1)
for timeout.After(time.Now()) {
   fmt.Println("doing work!")
   time.Sleep(time.Millisecond * 50)
}
```

The folowing docs may be helpful in this regard and include some examples:

- https://pkg.go.dev/time#Time.After
- https://pkg.go.dev/time#Time.Add

ccfishk replied:
Your suggestion of assert.Eventually is more concise solution in this case. 
What do you think ?

shaneutt replied:
Yes, I wrote this before I had thought on this. assert/require Eventually will both do the correct thing.

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,286 @@
+	for timeout < 60 {
+		podList, err := cluster.Client().CoreV1().Pods("knative-serving").List(ctx, metav1.ListOptions{})
+		if err != nil {
+			fmt.Println("failed retrieving knative pods.", err)
```
If you make this testing function accept a `testing.T` object you'll be able to use the `t.Log` methods to log these instead of printing directly to stdout, which will be helpful for making sure the logging output from this function is consistent with the rest of the test logging.

ccfishk replied:
Good suggestion, t.Log should be applied more. Thanks.

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,286 @@
+		}
+	}
+
+	fmt.Println("Failed to bring up knative resources within 60 seconds.")
```
It seems this entire function is meant to wait for knative resources to be ready, but creates it's own machinery for doing so. It seems to me that we should use `require.Eventually` for this purpose instead as this will be more consistent with the rest of the integration tests:

https://pkg.go.dev/github.com/stretchr/testify@v1.7.0/require#Eventually

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,286 @@
+}
+
+func TestKnativeIngress(t *testing.T) {
+	t.Skip("please use the steps run the test separately. ")
```
It seems like this may be a leftover from some debugging, should we remove this now?

```suggestion
```

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,286 @@
+	assert.NoError(t, err)
+	err = configKnativeDomain(ctx, proxy, cluster)
+	assert.NoError(t, err)
+	// cover webhook sync windows
+	time.Sleep(3 * time.Second)
```
This is technically a race condition, I recommend we use a `require.Eventually` with a specific API call which exercises the webhook to ensure it comes up before proceeding, but also fail here if the webhook does not provision properly.

ccfishk replied:
ATM, would say the knative webhook service readiness threshold is not stable - most cases ready for a post; sometimes, depends on the k8s cluster resources needs a window from configured to readiness. 
In our case,  I think a couple of seconds sync would be the most efficient way.

shaneutt replied:
I hope I can be persuasive here: the benefit of using `require.Eventually` is that you can perform some validating action and if this fails it will stop the entire test and report exactly where the problem occurred. The sleep method would be more prone to resulting in unknown behaviors: the later test calls will be written to expect that the hooks are working, but nothing actually validated that they did so the behavior of the tests past that point can be variable, resulting in some potential confusion under failure conditions. Let me know your thoughts?

ccfishk replied:
I hope following is clear: 
This unstable window after configuration before readiness is knative webhook service behavior, not Kong. As kong tests, we just use it.

Let me know what you think ?

shaneutt replied:
I'm still feeling pretty strongly that there's a more concise and definitive way we can test the "readiness" of the Knative webhook, rather than simply wait for a time period without validating that it's functional before we proceed to tests that assume they're functional. It seems like waiting, but then doing some kind of validating API call (like a dry run) would make for a stronger test.

ccfishk replied:
Agree.

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,286 @@
+	}
+
+	t.Log("clean up test deployments.")
+	deleteManifest(knativeCrds, ctx)
+	deleteManifest(knativeCore, ctx)
+	deleteManifest("helloworldgo.yaml", ctx)
```
I mentioned this earlier but I don't feel we should be dealing with manifests at all but instead the native Go types for consistency and for easier to track failures if an upgrade to a dependency breaks the API.

ccfishk replied:
agree that : don't need dealing with manifest at all since we tear down test cluster in the end. Also, the knative namespaces takes forever to get removed, which will significantly extend our CI time. Just drop them as it is.

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,286 @@
+	deleteManifest(knativeCrds, ctx)
+	deleteManifest(knativeCore, ctx)
+	deleteManifest("helloworldgo.yaml", ctx)
+	time.Sleep(5 * time.Second)
```
If this is settle time for teardowns, we could use a `require.Eventually` or `assert.Eventually` here to validate the teardown instead.

ccfishk replied:
kindly refer previous response.

shaneutt reviewed (changes requested):
Really excited to see this coming along so quickly, I have a few questions and comments for your considerations.

ccfishk reviewed (commented):
response

ccfishk reviewed (commented):
responds comments.

ccfishk commented:
@shaneutt  addressed your comments. The conversation within this PR has been long, most of them is about the knative integration test, I really think most of the suggestions are valuable and we can create issues on top of it, i.e add integration test dependencies into KTF etc. 

Kindly let me know what you think atm.

rainest commented:
This doesn't yet have changes to permissions and seems like it should, similar to https://github.com/Kong/kubernetes-ingress-controller/pull/1439/files#diff-2488970c7df4ebe27c81ac9edcd74832109f02ca5db967f36d46d1bfeaa45748

`make manifests` under the railgun directory should fix that.

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,257 @@
+}
+
+func TestKnativeIngress(t *testing.T) {
+	_ = proxyReady()
```
Instead of using `retrieveProxyInfo`, the URLs for the proxy Admin API and the Proxy URL itself are available in the return here, we should use these instead.

```suggestion
	p = proxyReady()
```

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,257 @@
+	}
+}
+
+func retrieveProxyInfo(ctx context.Context, t *testing.T) string {
+	var proxy string
+	assert.Eventually(t, func() bool {
+		cmd := exec.CommandContext(ctx, "kubectl", "get", "service", "ingress-controller-kong-proxy", "--namespace", "kong-system")
+		stdout, stderr := new(bytes.Buffer), new(bytes.Buffer)
+		cmd.Stdout = stdout
+		cmd.Stderr = stderr
+		if err := cmd.Run(); err != nil {
+			fmt.Fprintln(os.Stdout, stdout.String())
+			fmt.Fprintln(os.Stderr, stderr.String())
+			return false
+		}
+
+		if len(stdout.String()) > 0 {
+			info := strings.Split(stdout.String(), "\n")
+			proxy = strings.Fields(info[1])[3]
+			t.Logf("kong-proxy %s", proxy)
+			if checkIPAddress(proxy, t) == true {
+				return true
+			}
+		}
+		return false
+	}, 60*time.Second, 1*time.Second, true)
+	return proxy
+}
```
We should be able to use the proxyReady() function to do this check AND get the URLs
```suggestion
```

shaneutt commented on railgun/test/integration/knative_test.go:
```diff
@@ -0,0 +1,257 @@
+}
+
+func configKnativeNetwork(ctx context.Context, cluster kind.Cluster, t *testing.T) error {
+	payloadBytes := []byte("{\"data\": {\"ingress.class\": \"kong\"}}")
```
We should use the dynamic ingress.class value from the test suite here:

```suggestion
	payloadBytes := []byte(fmt.Sprintf("{\"data\": {\"ingress.class\": \"%s\"}}", ingressClass))
```

see: https://github.com/Kong/kubernetes-ingress-controller/blob/main/railgun/test/integration/suite_test.go#L54

shaneutt reviewed (changes requested):
Some additional comments, and it would seem we need to generate permissions and regenerate the controllers:

- `make manifests`
- `make generate.controllers`

shaneutt commented on railgun/manager/run.go:
```diff
@@ -261,6 +269,22 @@ func Run(ctx context.Context, c *config.Config) error {
+			},
+		}
+		controllers = append(controllers, controller)
+	}
```
We should add logging that informs the operator when this controller is (automatically) disabled.

ccfishk replied:
If user does not create and register controller using manager setup, the controller-runtime/manager does not include the specific operator during running. It is a built-in mechanism.

shaneutt replied:
I still think that we should be indicating to the user via logs that the controller was disabled if they tried to specifically enable it, however for the purposes of getting this PR completed I'm OK with skipping on this for now and following up later:

https://github.com/Kong/kubernetes-ingress-controller/issues/1448

shaneutt commented on railgun/manager/run.go:
```diff
@@ -286,6 +310,50 @@ func Run(ctx context.Context, c *config.Config) error {
 }
+
+// wait for knative cr before register and starting knative controller
+func FlipKnativeController(mgr manager.Manager, prx proxy.Proxy, enablestatus *util.EnablementStatus, cfg *config.Config, log logr.Logger) error {
```
Fundamentally I like what we're trying to do here: adding watchers which can proactively identify APIs that are added and hotload the controllers is a good feature.

However, I think we will need to do more to support this feature, the implementation so far has a few troubles and it also seems like something we're going to have to standardize further on. There's some growth of scope for KIC 2.0 which I don't feel we should be taking on just yet.

My recommendation:

1) add logging to the controller manager to inform the operator of decisions made to automatically disable the Knative Ingress controller
2) drop this function
3) create a follow up issue for us to discuss and groom which suggests this functionality for a separate scope

ccfishk replied:
This mechanism dynamically detect newly deployed CRD/CR after KIC is UP, and purposedly notify setup manager using registration and event setup.

 It is dedicated go routines that bump up controllers dynamically. we we can use the mechanism  dropping current pre-requisite CRD deployments such as KongClusterPlugin, KongConsumer, KongIngress, KongPlugin, TCPIngress, UDPIngress if there is restrict requirement from the cluster KIC runs on.

follow up issues for better log/convention/performance tests could be created for discussion,  they are awesome ideas.

shaneutt replied:
I don't feel like this was actually ready to be resolved as I still have some lingering concerns with the current implementation, however as this PR has been stuck in a "limbo" for some time I have created a follow up where we can look further into this so that it does not block current progress:

https://github.com/Kong/kubernetes-ingress-controller/issues/1449

shaneutt reviewed (changes requested):
A couple small comments: I think we're going to need to pull back on the background controller loader for knative.

shaneutt reviewed (approved):
I added a couple follow up items, but otherwise let's get the value we have here merged in! :+1:
